delivery_carrier_label_ups/models/stock_picking.py

```diff
@@ -0,0 +1,262 @@
+            # so we could also write it here immediately
+            result.append(
+                {
+                    "exact_price": price,
+                    "tracking_number": status["ShipmentResponse"]["ShipmentResults"][
+                        "ShipmentIdentificationNumber"
+                    ],
+                }
+            )
+        return result
+
+    def _ups_cancel(self):
+        """Cancel shipping"""
+        # note UPS also supports cancelling specific packages in a shipping,
+        # but that's not what we do here
+        for this in self:
+            status = this._ups_request(
+                "cancel", method="delete", url_parameters=(this.carrier_tracking_ref,)
+            )
+            this._ups_raise_error(status)
+
+    def _ups_shipping_data(self):
+        """Return a dict that can be passed to the shipping endpoint of the UPS API"""
+        self.ensure_one()
+        account = self._ups_auth_account()
+        return dict(
+            ShipmentRequest=dict(
+                Shipment=dict(
+                    Description=self._ups_reference(),
+                    Shipper=self._ups_partner_to_shipping_data(
+                        self.company_id.partner_id,
+                        ShipperNumber=account.ups_shipper_number,
+                    ),
+                    ShipTo=self._ups_partner_to_shipping_data(self.partner_id),
+                    ShipFrom=self._ups_partner_to_shipping_data(
+                        self.picking_type_id.warehouse_id.partner_id
+                    ),
+                    PaymentInformation=dict(
+                        ShipmentCharge=dict(
+                            Type="01",
+                            BillShipper=dict(
+                                # we ignore the alternatives paying per credit card or
+                                # paypal for now
+                                AccountNumber=account.ups_shipper_number,
+                            ),
+                        )
+                    ),
+                    Service=dict(
+                        Code=self.option_ids[:1].code,
+                        Description=self.option_ids[:1].name,
+                    ),
+                    Package=[
+                        self._ups_quant_package_data(package)
+                        for package in self.package_ids
+                    ],
+                ),
+            )
+        )
+
+    def _ups_partner_to_shipping_data(self, partner, **kwargs):
+        """Return a dict describing a partner for the shipping request"""
+        return dict(
+            **kwargs,
+            Name=(partner.parent_id or partner).name,
+            AttentionName=partner.name,
+            TaxIdentificationNumber=partner.vat,
+            Phone=dict(Number=partner.phone or partner.mobile,),
+            EMailAddress=partner.email,
+            Address=dict(
+                AddressLine=partner.street,
```

Comment:
We need to pass on street2 here as well. In the XML specs there is AddressLine1 and AddressLine2. I'm surprised the REST API has different field names.

Reply:
I also found out late that the REST API seems to be somehow derived from the XML API. At least we need to do atrocities like casting our numbers to strings for JSON...

This is the first time I need to pass multiple values for the same field, the wording of the documentation makes me think we can pass an array of strings here, that's what I do now. The API doesn't scream, but this is also something we need to check with the live URL because currently there's no way to see how UPS interprets whatever the result of a successful API call is.

Reply:
Did you manage to do this testing?

Reply:
Ping ;-)

Reply:
yes, I pushed orders to UPS live, printed labels from weird and non weird countries/addresses, chose different service options, and in the end cancelled my orders, worked fine

Reply:
Great, thanks!